rustls/src/server/mod.rs

```diff
@@ -324,113 +337,210 @@ impl ServerConnection {
-    }
+    /// Check if a `ClientHello` message has been received.
+    ///
+    /// Returns an error if the `ClientHello` message is invalid or if the acceptor has already
+    /// yielded an [`Accepted`]. Returns `Ok(None)` if no complete `ClientHello` has been received
+    /// yet.
+    pub fn accept(&mut self) -> Result<Option<Accepted<'_>>, Error> {
+        let mut connection = match self.inner.take() {
+            Some(conn) => conn,
+            None => {
+                return Err(Error::General(
+                    "cannot accept after successful acceptance".into(),
+                ));
+            }
+        };
 
-    fn reader(&mut self) -> Reader {
-        self.common.reader()
+        match connection.first_handshake_message() {
+            Ok(msg) => self.message = msg,
+            Err(e) => {
+                self.inner = Some(connection);
+                return Err(e);
+            }
+        }
+
+        let message = match self.message.as_ref() {
+            Some(msg) => msg,
+            None => {
+                self.inner = Some(connection);
+                return Ok(None);
+            }
+        };
+
+        let (client_hello, sig_schemes) = hs::process_client_hello(
+            message,
+            false,
+            &mut connection.common_state,
+            &mut connection.data,
+        )?;
+
+        Ok(Some(Accepted {
+            connection,
+            message,
+            client_hello,
+            sig_schemes,
+        }))
     }
 }
 
-impl PlaintextSink for ServerConnection {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        Ok(self.send_some_plaintext(buf))
-    }
+/// Represents a `ClientHello` message received through the [`Acceptor`].
+///
+/// Contains the state required to resume the connection through [`Accepted::into_connection()`].
+pub struct Accepted<'a> {
+    connection: ConnectionCommon<ServerConnectionData>,
+    message: &'a Message,
+    client_hello: &'a ClientHelloPayload,
+    sig_schemes: Vec<SignatureScheme>,
+}
 
-    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> io::Result<usize> {
-        let mut sz = 0;
-        for buf in bufs {
-            sz += self.send_some_plaintext(buf);
-        }
-        Ok(sz)
+impl<'a> Accepted<'a> {
+    /// Get the [`ClientHello`] for this connection.
+    pub fn client_hello(&self) -> ClientHello<'_> {
+        ClientHello::new(
+            &self.connection.data.sni,
+            &self.sig_schemes,
+            self.client_hello.get_alpn_extension(),
+        )
     }
 
-    fn flush(&mut self) -> io::Result<()> {
-        Ok(())
+    /// Convert the [`Accepted`] into a [`ServerConnection`].
+    ///
+    /// Takes the state returned from [`Acceptor::accept()`] as well as the [`ServerConfig`] and
+    /// [`sign::CertifiedKey`] that should be used for the session. Returns an error if
+    /// configuration-dependent validation of the received `ClientHello` message fails.
+    pub fn into_connection(
```

Comment:
While the type-state things are neat from the correctness point of view they also come with a cost:

Now the whole state has to be transferred into another location. If the state is bigger, it might be a more costly memcpy on the stack. Also for users, if they `Box`ed the `Connection`  before, they might now need a a separate `Box`  (or an enum) to do the transformation.

If the connection state is not that big it's likely not an issue. But I would recommend doing some benchmarking (e.g. by just doing handshakes in row) to determine the impact.

Also is the plan to go further and add a separate typestate to differentiate between the connection where the handshake is still in progress vs completed?

Reply:
Most of the stuff in `ConnectionCommon` is boxed up anyway (the state machine state, the buffers), so I don't think the type itself is particularly large. I don't know what you mean about boxing the `Connection` before -- there is no `Connection`, so you can't box it.

I don't currently have plans to differentiate between handshake in progress vs completed for the short term at least.

Reply:
> I don't know what you mean about boxing the Connection before 

I meant `Acceptor`. Good to know most things are boxed here